### ISSUE TITLE
关于VisDrone2019数据集训练的精度问题

### ISSUE THREAD

**Issue by zgh2022** [Peterande/D-FINE#37]
你好，我正在用readme中提到的非常推荐的Objects365 pre-trained model for fine-tuning方式去微调visdrone数据集，用的x模型，但是loss一直维持在40左右，而map也是非常的低，这是训练了30多轮的结果，以及我的参数设置：
![Screenshot from 2024-11-08 09-05-27](https://github.com/user-attachments/assets/295415c7-3ff3-4a8d-9883-25984fd00250)
![Screenshot from 2024-11-08 09-04-55](https://github.com/user-attachments/assets/e4b924c4-4f63-4944-bd32-ffe9e3ceeb36)
我的数据集有经过yolo2coco的转换，但是我可视化后发现数据集是没有问题的，我想知道这种情况是否正常

**Comment by zgh2022** [comment]
对了，训练环境是单卡3090

**Comment by Peterande** [comment]
Precision非常非常低，大概率还是数据的问题。最好和原始COCO数据集进行一下对比，或者用一些其他COCO形式的数据集在您的环境训练看看是否正常

**Comment by zgh2022** [comment]
您好，我想我找到部分问题了，我使用了一个非常小的数据集（coco格式）https://universe.roboflow.com/johannes-kroger-t7hyr/mulleimererkennung/dataset/1去进行了测试，然而结果是：
![Screenshot from 2024-11-08 14-37-49](https://github.com/user-attachments/assets/f0c86595-bc04-46fa-a8f3-dba2676839e8)
但是当我根据文档修改了solver里的obj365_ids后：
![Screenshot from 2024-11-08 14-38-04](https://github.com/user-attachments/assets/3dcfa6aa-2fe6-443f-b334-0ba8dcd2087d)
map变成了：
![Screenshot from 2024-11-08 14-42-23](https://github.com/user-attachments/assets/aa67ff11-f6dc-4942-af53-67440b268b89)
但是相关指标实际上也就固定在这些数值了
所以是否某个类不在 self.obj365_ids 列表中时，那么对应的权重可能不会被正确地初始化或调整？是否在评估过程中，模型只会关注 self.obj365_ids 列表中的类别呢？可能是map_class_weights 方法的处理问题吗？
另外我尝试了将ids改成了与bin大致相同的bottle类和完全不同的其他类，评估结果也确实如此：
![Screenshot from 2024-11-08 15-01-49](https://github.com/user-attachments/assets/59d01ea9-34b4-4b45-99a5-16466cf07cb9)

**Comment by Peterande** [comment]
根据https://colab.research.google.com/github/Kazuhito00/D-FINE-Colaboratory-Training-Sample/blob/main/D-FINE-Colaboratory-Training-Sample.ipynb
非常小的数据集要到很后期才能有AP的涨幅，前期确实非常接近0。
另外你可以尝试注释下面这行，这代表不进行任何class mapping:
https://github.com/Peterande/D-FINE/blob/9f8c2b430f6e9d40aac966b79984cc2c9f88f39b/src/solver/_solver.py#L181

**Comment by zgh2022** [comment]
我使用visdrone数据集训练了180轮，这是训练日志和控制台最后20轮的输出：
[log.txt](https://github.com/user-attachments/files/17693999/log.txt)
[result.txt](https://github.com/user-attachments/files/17694004/result.txt)
所以我在想是不是微调模式下的AP计算的部分有点问题呢？

**Comment by Peterande** [comment]
> 我使用visdrone数据集训练了180轮，这是训练日志和控制台最后20轮的输出： [log.txt](https://github.com/user-attachments/files/17693999/log.txt) [result.txt](https://github.com/user-attachments/files/17694004/result.txt) 所以我在想是不是微调模式下的AP计算的部分有点问题呢？

AP计算代码和是否微调没有关系，微调的区别只在于权重的初始化

**Comment by zgh2022** [comment]
这个我知道，我指的是我修改了各种参数以及包括不进行任何class mapping之后，使用visdrone数据集训练的过程中仍然出现了AP的计算没有提升这一现象，可以看终端输出的result.txt文件，AP基本保持在那个数值就没有提升了，这会带来一个问题就是无法获得best模型，因为我在训练结束后使用了best模型和last模型分别进行了检测，last模型的效果是完全优于best模型的，所以我才会去想是否关于自定义的数据集的AP计算的地方有什么我不知道的点需要去修改（除了remap_mscoco_category）

**Comment by Peterande** [comment]
现在的代码版本前期保存的是best_stg1.pth。是不是你的代码中有两个版本冲突的情况

**Comment by S130111** [comment]
我也有同样的问题 请问解决了吗

**Comment by ZYHQQ-666** [comment]
@S130111 兄弟，请问你解决这个问题了吗

**Comment by RicePasteM** [comment]
我抽空来复现一下这个问题。

**Comment by RicePasteM** [comment]
感谢反馈。我们在 VisDrone2019-DET 上重新进行了复现。
使用正确转换并校验过的 COCO 标注、D-FINE-X Objects365 权重以及 issue 中的训练参数时，前几轮 AP 接近 0，但随后可以正常收敛：
- Epoch 9：AP 11.62%，AP50 18.58%
- 扩展训练最佳：AP 26.32%，AP50 42.97%

因此，issue 中训练 180 个 epoch 后 AP 仍停留在约 3.5% 的现象无法在正确标注上复现，问题大概率来自 YOLO→COCO 的转换结果，而不是学习率或 AP 计算代码。

另外，Objects365 权重无法匹配自定义的 10 类分类头，因此分类头会随机初始化，前几轮 AP 很低属于正常现象。总 loss 也是多个 decoder、auxiliary 和 denoising loss 的总和，数值在 30～40 并不代表训练异常。

建议重点检查：

1. `category_id` 是否连续且从 0 开始；
2. COCO bbox 是否为绝对像素格式 `[x, y, width, height]`；
3. 是否正确过滤 VisDrone 中 `score=0`、类别 0 和类别 11 的标注；
4. 标注引用的图片尺寸和文件名是否正确。

如果仍有问题，请提供实际训练使用的 COCO JSON，我们可以进一步定位具体的转换错误。

我们额外提供了数据集转换的工具脚本，希望能对你产生帮助：[convert_visdrone_to_coco.py](https://github.com/user-attachments/files/30885261/convert_visdrone_to_coco.py)

**Comment by ZYHQQ-666** [comment]
<img width="910" height="106" alt="Image" src="https://github.com/user-attachments/assets/6b077799-825b-47c2-b908-736158793c3b" />成了，三个epo就有18
